app/services/market_analysis_core.py

```
@@ -784,13 +838,25 @@ async def _collect_symbol_data(
+                return None
             try:
                 price_info = await asyncio.wait_for(
                     self._get_symbol_price(exchange, symbol),
                     timeout=self._per_exchange_timeout,
                 )
-                return {"exchange": exchange, **price_info} if price_info else None
+                if price_info:
+                    await self._record_exchange_success(exchange)
+                    return {"exchange": exchange, **price_info}
+                await self._record_exchange_failure(exchange)
+                return None
```

Ruff (0.14.3)

Consider moving this statement to an `else` block

(TRY300)